policies/ingress/unique_ingress_host.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2019 Cruise LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//    https://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.package ingress
+
+package ingress
+
+import (
+	"context"
+	"flag"
+
+	"github.com/cruise-automation/k-rail/policies"
+	"github.com/cruise-automation/k-rail/resource"
+	log "github.com/sirupsen/logrus"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func NewPolicyRequireUniqueHost() (PolicyRequireUniqueHost, error) {
+	p := PolicyRequireUniqueHost{}
+
+	var kubeconfig = flag.String("kubeConfig", "", "absolute path to the kubeconfig file: `<home>/.kube/config`")
+	flag.Parse()
+	flag.Set("logtostderr", "true")
+
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		return p, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return p, err
+	}
+	p.client = clientset
+	return p, err
+}
+
+type PolicyRequireUniqueHost struct {
+	client *kubernetes.Clientset
+}
+
+func (p PolicyRequireUniqueHost) Name() string {
+	return "ingress_unique_ingress_host"
+}
+
+func (p PolicyRequireUniqueHost) CheckIngressNamespaces(ctx context.Context, host string) ([]string, error) {
+	ingressNamespaces := []string{}
+	ingresses, err := p.client.ExtensionsV1beta1().Ingresses("").List(metav1.ListOptions{})
+	if err != nil {
+		return ingressNamespaces, err
+	}
+	for _, ingress := range ingresses.Items {
+		rules := ingress.Spec.Rules
+		for _, rule := range rules {
+			if rule.Host == host {
+				ingressNamespace := ingress.ObjectMeta.Namespace
+				ingressNamespaces = append(ingressNamespaces, ingressNamespace)
+			}
+		}
+	}
+	return ingressNamespaces, nil
+}
+
+func Find(slice []string, val string) bool {
```

Comment:
This should be unexported so we don't pollute the package. Also, with larger amounts of data (which we don't have), the most efficient way to do this is a map[string]struct{}, because you can then do hash map lookups instead of iterating through all elements.

```
namespaces := map[string]struct{}{}
namespaces["some-namespace"] = struct{}{}  // struct{} is an empty object that takes little memory
_, ok := namespace["some-other-namespace"]
if ok {
  // do something
}
```